src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/IDE/VSSolutionManager.cs

```diff
@@ -342,19 +363,43 @@ public bool IsSolutionOpen
         {
             get
             {
-                return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+                if (_solutionEventsCookie != 0)
                 {
-                    return await IsSolutionOpenAsync();
-                });
+                    return _isSolutionOpen;
+                }
+                else
+                {
+                    return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
+                    {
+                        return await IsSolutionOpenAsync();
+                    });
+                }
             }
         }
 
         public async Task<bool> IsSolutionOpenAsync()
         {
-            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
-
-            var vsSolution = await _asyncVSSolution.GetValueAsync();
-            return IsSolutionOpenFromVSSolution(vsSolution);
+            if (_solutionEventsCookie != 0)
+            {
+                return _isSolutionOpen;
+            }
+            else
+            {
+                if (NuGetUIThreadHelper.JoinableTaskFactory.Context.IsOnMainThread)
+                {
+                    await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var vsSolution = await _asyncVSSolution.GetValueAsync();
+                    return IsSolutionOpenFromVSSolution(vsSolution);
+                }
+                else
+                {
+                    var vsBackgroundSolution = await _asyncVSBackgroundSolution.GetValueAsync();
+                    // VS SDK needs to be updated to not give this incorrect advice
+#pragma warning disable VSTHRD010 // Invoke single-threaded types on Main thread
+                    return vsBackgroundSolution.SolutionDirectory != null;
```

Comment:
The analyzers allow us to specify additional methods that are free threaded. Worth looking into that.

Reply:
As mentioned in another thread, I'm removing usage of `IVsBackgroundSolution` for now because it's going to change namespaces, and I'm worried about it being a breaking change.